warehouse/accounts/views.py

```diff
@@ -122,38 +124,113 @@ def login(request, redirect_field_name=REDIRECT_FIELD_NAME, _form_class=LoginFor
             username = form.username.data
             userid = user_service.find_userid(username)
 
+            # If the user has enabled two factor authentication.
+            if user_service.has_two_factor(userid):
+                two_factor_data = {"userid": userid}
+                if redirect_to:
+                    two_factor_data["redirect_to"] = redirect_to
+
+                token_service = request.find_service(ITokenService, name="two_factor")
+                token = token_service.dumps(two_factor_data)
+
+                # Stuff our token in the query and redirect to two-factor page.
+                resp = HTTPSeeOther(
+                    request.route_path("accounts.two-factor", _query=token)
+                )
+
+                return resp
+            else:
+                # If the user-originating redirection url is not safe, then
+                # redirect to the index instead.
+                if not redirect_to or not is_safe_url(
+                    url=redirect_to, host=request.host
+                ):
+                    redirect_to = request.route_path("manage.projects")
+
+                # Actually perform the login routine for our user.
+                headers = _login_user(request, userid)
+
+                # Now that we're logged in we'll want to redirect the user to
+                # either where they were trying to go originally, or to the default
+                # view.
+                resp = HTTPSeeOther(redirect_to, headers=dict(headers))
+
+                # We'll use this cookie so that client side javascript can
+                # Determine the actual user ID (not username, user ID). This is
+                # *not* a security sensitive context and it *MUST* not be used
+                # where security matters.
+                #
+                # We'll also hash this value just to avoid leaking the actual User
+                # IDs here, even though it really shouldn't matter.
+                resp.set_cookie(
+                    USER_ID_INSECURE_COOKIE,
+                    hashlib.blake2b(str(userid).encode(), person=b"warehouse.userid")
```

Comment:
The diff is a little hard to read here, but this was changed from being explicitly ASCII-encoded to using the default encoding. It's not clear to me why?

Reply:
This PR mostly uses `encode()` since the default arg is `utf8` (which works fine here of course). However, a quick grep of the warehouse codebase appears to show a preference for explicit encoding arguments so it should probably be changed across this whole PR. What do you think @woodruffw?

Reply:
Looks like this got caught up in some churn during the TOTP cookie changes. It shouldn't affect the correctness of the `set_cookie`, but I'll roll it back since it's vestigial.

Reply:
That's what I get for commenting without refreshing! I agree with @reaperhulk, will also add explicit encodings to the rest of this changeset.